app/controllers/api/v0/highlights_controller.rb

```diff
@@ -9,7 +9,9 @@ def create
     inbound_binding, error = bind(params.require(:highlight), Api::V0::Bindings::NewHighlight)
     render(json: error, status: error.status_code) and return if error
 
-    model = inbound_binding.create_model!(user_uuid: current_user_uuid)
+    ServiceLimits.create_check(current_user_uuid, inbound_binding) do |binding|
+      model = binding.create_model!(user_id: current_user_uuid)
+    end
```

Comment:
Until I realized that the `validate_blah` methods inside `ServiceLimits` were updating fields themselves (which need a transaction for rollback behavior), it wasn't clear to me why the `create_model!` call was being passed in in a block.  Seemed like the `ServiceLimits` call could have just been a line on its own.

So I'm wondering if some renaming/rephrasing is warranted.  E.g. it can be framed more as a guard around the create call:

```ruby
model = ServiceLimits.with_create_protection do
  inbound_binding.create_model!(user_id: current_user_uuid)
end
```

```ruby
class ServiceLimits
  def self.with_create_protection
    yield.tap do |model| # more fun ways of using `tap` ;-)
      track_and_validate_max_highlights(model) 
      track_and_validate_max_annotation_chars(model)
    end
  end
end
```

By creating the model right away with the `yield` call, you don't need to pass args in to `with_create_protection` that you are also using when you create the model, you can just pull the data from the created model. This lets you not repeat `user = User.find_or_create_by(id: user_uuid)` and do less passing around of the binding which I see as good because the binding is API-specific but the `ServiceLimits` class is not.

I also like `track_and` in the private method names because it makes it more clear that the method is updating records (more than just validating values).

Your method as currently written as well as what I just wrote above also suffer from an assumption that the line that creates the model is using `create!` instead of `create`.  If someone didn't use the exception-raising `create!` variety, both of our implementations would return unsaved models that we'd then use to update tracked service limit values.  My implementation would be improved by:

```ruby
class ServiceLimits
  def self.with_create_protection
    model = yield # if this is using `create!` and there's an error, the exception will cause us to bail here

    # however if it was just a `create` call we need other checks
    raise ArgumentError, "Block did not yield an active record model" unless model.is_a?(Highlight)
    raise <something else> if model.errors?

    track_and_validate_max_highlights(model) 
    track_and_validate_max_annotation_chars(model)

    model
end
```